src/crypto/ecops64/ecops64-c.c

```diff
@@ -0,0 +1,489 @@
+	int carry, carry2;
+
+	carry = 0; /* 0..1 */
+	for (uint64_t i = 0; i < 31; i++)
+	{
+		carry += s[i]; /* 0..256 */
+		carry2 = (carry + 8) >> 4; /* 0..16 */
+		b[2 * i] = carry - (carry2 << 4); /* -8..7 */
+		carry = (carry2 + 8) >> 4; /* 0..1 */
+		b[2 * i + 1] = carry2 - (carry << 4); /* -8..7 */
+	}
+
+	carry += s[31]; /* 0..128 */
+	carry2 = (carry + 8) >> 4; /* 0..8 */
+	b[62] = carry - (carry2 << 4); /* -8..7 */
+	b[63] = carry2; /* 0..8 */
+}
+
+/* Multiples of the base point in Niels' representation */
+extern const ge64_niels ge64_base_multiples_niels[];
+extern const fe64 ge64_base_mult0_x[];
+extern const fe64 ge64_base_mult0_y[];
+extern const fe64 ge64_base_mult0_t[];
+
+void ge64_scalarmult_base(ge64_p3 *r, const unsigned char s[32])
+{
+	signed char b[64];
+	mult_extract(b, s);
+
+	fe64_copy(r->x, ge64_base_mult0_x[b[0] + 8]);
+	fe64_copy(r->y, ge64_base_mult0_y[b[0] + 8]);
+	fe64_copy(r->t, ge64_base_mult0_t[b[0] + 8]);
+	fe64_setint(r->z, 2);
+
+	for(uint64_t i=1;i<64;i++)
+	{
+		uint64_t idx = (i-1) * 17 + b[i] + 8;
+		ge64_nielsadd2(r, &ge64_base_multiples_niels[idx]);
+	}
+}
+
+static inline void ge64_p3_to_pniels(ge64_pniels *r, const ge64_p3 *p)
+{
+	fe64_add(r->xaddy, p->x, p->y);
+	fe64_sub(r->ysubx, p->y, p->x);
+	fe64_copy(r->z, p->z);
+	fe64_mul(r->t2d, p->t, fe64_2d);
+}
+
+/* Assumes that a[31] <= 127 */
+void ge64_scalarmult(ge64_p2 *r, const unsigned char a[32], const ge64_p3 *A)
+{
+	signed char e[64];
+	mult_extract(e, a);
+
+	ge64_pniels Ai[17];
+
+	fe64_setint(Ai[8].xaddy, 1);
+	fe64_setint(Ai[8].ysubx, 1);
+	fe64_setint(Ai[8].z, 1);
+	fe64_setint(Ai[8].t2d, 0);
+
+	ge64_p3_to_pniels(&Ai[9], A);
+
+	ge64_p3 u;
+	ge64_p1p1 t;
+	for (uint64_t i = 9; i < 16; i++)
+	{
+		ge64_pnielsadd_p1p1(&t, A, &Ai[i]);
+		ge64_p1p1_to_p3(&u, &t);
+		ge64_p3_to_pniels(&Ai[i + 1], &u);
+	}
+
+	for (uint64_t i = 0; i < 8; i++)
+	{
+		fe64_copy(Ai[i].xaddy, Ai[16-i].ysubx);
+		fe64_copy(Ai[i].ysubx, Ai[16-i].xaddy);
+		fe64_copy(Ai[i].z, Ai[16-i].z);
+		fe64_neg(Ai[i].t2d, Ai[16-i].t2d);
+	}
+
+	fe64_setint(r->x, 0);
+	fe64_setint(r->y, 1);
+	fe64_setint(r->z, 1);
+
+	for (int64_t i = 63; i >= 0; i--)
+	{
+		ge64_p2_dbl(&t, r);
+		ge64_p1p1_to_p2(r, &t);
+		ge64_p2_dbl(&t, r);
+		ge64_p1p1_to_p2(r, &t);
+		ge64_p2_dbl(&t, r);
+		ge64_p1p1_to_p2(r, &t);
+		ge64_p2_dbl(&t, r);
+		ge64_p1p1_to_p3(&u, &t);
+
+		uint64_t idx = e[i] + 8;
+		ge64_pnielsadd_p1p1(&t, &u, &Ai[idx]);
+		ge64_p1p1_to_p2(r, &t);
+	}
+}
+
+inline void ge64_sub(ge64_p1p1 *r, const ge64_p3 *p, const ge64_p3 *q) 
+{
+	fe64 t0, xaddy, ysubx, t2d;
+	fe64_add(xaddy, q->x, q->y);
+	fe64_sub(ysubx, q->y, q->x);
+	fe64_mul(t2d, q->t, fe64_2d);
+	
+	fe64_add(r->x, p->y, p->x);
+	fe64_sub(r->y, p->y, p->x);
+	fe64_mul(r->z, r->x, ysubx);
+	fe64_mul(r->y, r->y, xaddy);
+	fe64_mul(r->t, t2d, p->t);
+	fe64_mul(r->x, p->z, q->z);
+	fe64_add(t0, r->x, r->x);
+	fe64_sub(r->x, r->z, r->y);
+	fe64_add(r->y, r->z, r->y);
+	fe64_sub(r->z, t0, r->t);
+	fe64_add(r->t, t0, r->t);
+	}
```

Comment:
Broken whitespace

Reply:
will update, I have not touched this file expect the license